src/Package.zig

```diff
@@ -327,30 +327,47 @@ pub fn fetchAndAddDependencies(
             name,
         );
 
-        if (!sub.found_existing) {
-            try sub.mod.fetchAndAddDependencies(
-                deps_pkg,
-                arena,
-                thread_pool,
-                http_client,
-                sub.mod.root_src_directory,
-                global_cache_directory,
-                local_cache_directory,
-                dependencies_source,
-                error_bundle,
-                all_modules,
-                root_prog_node,
-                dep.hash.?,
-            );
-        }
+        if (sub.mod) |mod| {
+            if (!sub.found_existing) {
+                try mod.fetchAndAddDependencies(
+                    deps_pkg,
+                    arena,
+                    thread_pool,
+                    http_client,
+                    mod.root_src_directory,
+                    global_cache_directory,
+                    local_cache_directory,
+                    dependencies_source,
+                    error_bundle,
+                    all_modules,
+                    root_prog_node,
+                    dep.hash.?,
+                );
+            }
 
-        try pkg.add(gpa, name, sub.mod);
-        if (deps_pkg.table.get(dep.hash.?)) |other_sub| {
-            // This should be the same package (and hence module) since it's the same hash
-            // TODO: dedup multiple versions of the same package
-            assert(other_sub == sub.mod);
-        } else {
-            try deps_pkg.add(gpa, dep.hash.?, sub.mod);
+            try pkg.add(gpa, name, mod);
+            if (deps_pkg.table.get(dep.hash.?)) |other_sub| {
+                // This should be the same package (and hence module) since it's the same hash
+                // TODO: dedup multiple versions of the same package
+                assert(other_sub == mod);
+            } else {
+                try deps_pkg.add(gpa, dep.hash.?, mod);
+            }
+        } else if (!sub.found_existing) {
+            const pkg_dir_sub_path = "p" ++ fs.path.sep_str ++ (dep.hash.?)[0..hex_multihash_len];
+            const build_root = try global_cache_directory.join(arena, &.{pkg_dir_sub_path});
+            try dependencies_source.writer().print(
+                \\    pub const {} = struct {{
+                \\        pub const build_root = "{}";
+                \\        pub const build_zig = void;
```

Comment:
nit: `void` is a slightly strange choice here. IMO the best choice is to just omit the decl altogether, and use `@hasDecl` in the build system (changing `dependencyInner`'s arg to `comptime build_zig: ?type`).